### DISCUSSION TITLE
thanhle.blog/frontend-performance-pattern-vi

### DISCUSSION THREAD

**Discussion by giscus** [thanhlmm/blog#51] [Ideas]
# thanhle.blog/frontend-performance-pattern-vi

Các pattern để tối ưu tốc độ của web app

https://thanhle.blog/blog/frontend-performance-pattern-vi

**Comment by dfjerry** [comment]
Không có nút vote nhỉ để em cho +n luôn, cứ gọi là bị cuốn.

**Reply by thanhlmm** [reply]
<img width="1167" alt="Screen Shot 2022-08-03 at 6 00 07 PM" src="https://user-images.githubusercontent.com/9281080/182592592-5b2f2c3b-6bd2-4ea7-9924-71df684301c6.png">

Haha thanks em. Có nút reaction nè. Nhớ subscribe để nhận email khi có bài mới nha 🤟

**Comment by viettl** [comment]
Hi anh, ở trên em thấy có nhắc đến việc anh đã handle nhiều input để improve performance, không biết nếu tiện thì anh có thể chia sẻ thêm về case đấy được không ạ, thanks anh hehe

**Reply by thanhlmm** [reply]
Hi, ý em muốn hỏi là `Debounce user input`?

**Reply by viettl** [reply]
dạ đoạn này anh ạ [Link to highlight](https://thanhle.blog/blog/frontend-performance-pattern-vi#:~:text=C%C5%A9ng%20c%C3%B3%20l%E1%BA%A7n%20m%C3%ACnh%20ph%E1%BA%A3i%20s%E1%BB%ADa%20l%E1%BA%A1i%20m%E1%BB%99t%20flow%20logic%20react%20m%C3%A0%20c%C3%B3%20form%20kho%E1%BA%A3ng%20g%E1%BA%A7n%2020%20fields%2C%20v%C3%A0%20b%E1%BA%A1n%20dev%20d%C3%B9ng%20useMemo%20cho%20t%E1%BA%A5t%20c%E1%BA%A3%20c%C3%A1c%20state%20%C4%91%C3%B3%2C%20d%E1%BA%ABn%20t%E1%BB%9Bi%20bug%20logic%20khi%20ch%E1%BA%A1y%20m%C3%A0%20%C4%91%E1%BA%BFn%20b%E1%BA%A1n%20%C4%91%C3%B3%20c%C5%A9ng%20kh%C3%B4ng%20bi%E1%BA%BFt%20s%E1%BB%ADa%20l%C3%A0m%20sao%20m%C3%A0%20ph%E1%BA%A3i%20nh%E1%BB%9D%20%C4%91%E1%BA%BFn%20free%20lancer%20%C4%91%E1%BB%83%20debug%20gi%C3%BAp.%20Sui%20r%E1%BB%A7i%20sao%20m%C3%ACnh%20l%C3%A0%20th%E1%BA%B1ng%20free%20lancer%20%C4%91%C3%B3!)

**Reply by thanhlmm** [reply]
À case đó cụ thể của mình là cái form có khoảng hơn 20 input, tất cả đều dùng `useState`. Chuyện sẽ bình thường cho tới khi bạn cứ `useMemo` cho tất cả các state bên trong đó, tuy nhiên việc code cứ update liên tục, dẫn tới list dependencies cho `useMemo` bị out date. Một thời gian thấy ngập một đống bug là code state đổi mà không render theo ý muốn.
Cách fix của anh lúc đó là xóa hết `useMemo` đi thôi.

Không biết em đang gặp case giống vậy hay không nhỉ?

**Reply by viettl** [reply]
Vâng, case này hay quá ạ, thanks anh đã share lại ạ :raised_hands:. 
Em thấy cũng nhiều người lạm dụng useCallback và useMemo quá. 
Hiện tại thì chưa gặp case đó ạ, em thấy thú vị nên xin anh share lại ấy.

**Reply by thanhlmm** [reply]
Uhm case này là do bạn focus vào performance quá nhiều nên dẫn tới bug cơ bản feature chạy ko được luôn.
